File: engine/src/invana/apps/govern/addressing.py
"""The address grammar — `<layer>/<sublayer>/<name>`, and the patterns that match it.

**The address is the only identifier**
([GV4](docs/for-developers/modules/govern/spec.md)): the lens matches on it, the
ledger records it, the run dashboard expands it. One string instead of five
shapes of rule, which is why this module is thirty lines of matching and not a
resolver per layer.

Two wildcards, and they are not interchangeable:

``*``
    exactly one segment. ``llm/anthropic-prod/*`` is every model on that
    provider row and nothing on another.
``**``
    the rest, including nothing. ``third_party/**`` is the whole layer, and
    ``third_party/api/clearbit.com/**`` is that host's every path.

A ``*`` may also appear **inside** a segment, which is how a rule names a model
across its versions: ``graph_data/model/Deals@*`` is every published version of
``Deals``, and writing the version out would make a bound that silently stops
applying the next time somebody publishes.

``**`` is only meaningful as the final segment; anywhere else it would make one
pattern match two different shapes of thing, and a rule nobody can read the
reach of is not a bound.
"""

from __future__ import annotations

import enum
import re
from dataclasses import dataclass
from fnmatch import fnmatchcase

#: Segments are the readable characters an address is built from. Deliberately
#: permissive about ``.``, ``@`` and ``:`` — ``Routes@v4``, ``clearbit.com`` and
#: ``claude-haiku-4.5`` are all one segment each.
_SEGMENT = re.compile(r"^[A-Za-z0-9_.@:+~-]+$")
#: The same, with ``*`` allowed as a glob within the segment.
_SEGMENT_GLOB = re.compile(r"^[A-Za-z0-9_.@:+~*-]+$")


class Layer(enum.StrEnum):
    """The five governed classes of participant, plus the spine.

    ``agent`` is the **spine** — the runtime itself, which does the
    participating rather than being a participant
    ([govern § 1](docs/for-developers/modules/govern/spec.md)). It is addressed
    and recorded like the rest so the ledger has one shape, and it is never
    governed: a rule that denies it would deny the run its own dispatches.
    """

    graph_data = "graph_data"
    llm = "llm"
    third_party = "third_party"
    cache = "cache"
    human = "human"
    agent = "agent"


#: The layers a rule may govern. The spine is not one of them.
GOVERNED_LAYERS: tuple[Layer, ...] = (
    Layer.graph_data,
    Layer.llm,
    Layer.third_party,
    Layer.cache,
    Layer.human,
)

#: What a layer's second segment may be. A sublayer is not a second vocabulary —
#: it is the touch record's ``kind`` (govern § 1). ``llm`` is the open one: its
#: sublayer is the *configured provider row*, whose name a person chose
#: ([GV9](docs/for-developers/modules/govern/spec.md)), so it cannot be a fixed
#: list here.
SUBLAYERS: dict[Layer, tuple[str, ...] | None] = {
    Layer.graph_data: ("model", "stitch", "dataset"),
    Layer.llm: None,
    Layer.third_party: ("api", "app", "db", "agent"),
    Layer.cache: ("answer", "prefix", "result"),
    Layer.human: ("person", "role"),
    Layer.agent: ("agent",),
}


@dataclass(frozen=True, slots=True)
class Address:
    """One participant, split into the three parts every surface reads back."""

    layer: Layer
    sublayer: str
    #: Everything after the sublayer, ``/`` included —
    #: ``clearbit.com/v2/companies`` is one name, not three segments.
    name: str

    def __str__(self) -> str:
        return f"{self.layer.value}/{self.sublayer}/{self.name}"


class AddressError(ValueError):
    """The string is not an address. Raised where a person typed one, so the
    message names which part is wrong rather than echoing the whole string."""


def parse_address(raw: str) -> Address:
    """``graph_data/model/Routes@v4`` → an :class:`Address`.

    Rejects wildcards: this parses a *participant*, and a participant with a
    ``*`` in it is a pattern that got as far as somewhere only concrete
    addresses belong.
    """
    parts = raw.strip().split("/")
    if len(parts) < 3 or not all(parts):
        raise AddressError("An address is <layer>/<sublayer>/<name>.")
    try:
        layer = Layer(parts[0])
    except ValueError:
        raise AddressError(f"Unknown layer {parts[0]!r}.") from None

    sublayer, name = parts[1], "/".join(parts[2:])
    allowed = SUBLAYERS[layer]
    if allowed is not None and sublayer not in allowed:
        raise AddressError(f"{layer.value} has no sublayer {sublayer!r} — one of {', '.join(allowed)}.")
    for segment in (sublayer, *name.split("/")):
        if "*" in segment:
            raise AddressError("An address names one participant; a pattern belongs on a rule.")
        if not _SEGMENT.match(segment):
            raise AddressError(f"{segment!r} is not a readable address segment.")
    return Address(layer=layer, sublayer=sublayer, name=name)


def validate_pattern(raw: str) -> None:
    """Raise :class:`AddressError` unless ``raw`` is a usable rule pattern.

    A pattern names a layer first — a rule that begins ``**`` would span every
    layer at once, and the five layers are the thing a person reasons about.
    """
    parts = raw.strip().split("/")
    if not parts or not all(parts):
        raise AddressError("A rule matches <layer>/<sublayer>/<name>, with * or ** for the parts it does not name.")
    try:
        Layer(parts[0])
    except ValueError:
        raise AddressError(f"A rule names a layer first; {parts[0]!r} is not one.") from None
    if "**" in parts[:-1]:
        raise AddressError("** matches the rest, so it can only be the last part of a pattern.")
    for segment in parts[1:]:
        if segment in ("*", "**"):
            continue
        if not _SEGMENT_GLOB.match(segment):
            raise AddressError(f"{segment!r} is not a readable address segment.")


def pattern_layer(raw: str) -> Layer:
    """The layer a pattern governs. Validated first, so this cannot fail after."""
    validate_pattern(raw)
    return Layer(raw.strip().split("/")[0])


def matches(pattern: str, address: str) -> bool:
    """Does ``pattern`` reach ``address``?

    Segment-wise, with ``*`` taking one and ``**`` taking the rest. Nothing here
    is about precedence: a broader pattern and a narrower one both simply match,
    and which one *wins* is
    [GV5](docs/for-developers/modules/govern/spec.md)'s answer — deny, at any
    specificity — decided by the caller and never by how long a pattern is.
    """
    pat = pattern.strip().split("/")
    addr = address.strip().split("/")

    for i, segment in enumerate(pat):
        if segment == "**":
            return True
        if i >= len(addr):
            return False
        if not _segment_matches(segment, addr[i]):
            return False
    return len(pat) == len(addr)


def _segment_matches(pattern: str, segment: str) -> bool:
    """One pattern segment against one address segment.

    ``*`` alone takes the whole segment; a ``*`` inside one is a glob over it,
    so ``Deals@*`` reaches ``Deals@1.0.0`` and ``Deal*`` would reach both
    ``Deals`` and ``Dealers`` — which is why a rule is written against the
    catalogue's live match preview rather than from memory.
    """
    if pattern == "*":
        return True
    if "*" not in pattern:
        return pattern == segment
    return fnmatchcase(segment, pattern)


def specificity(pattern: str) -> int:
    """How precisely a pattern names something — concrete segments, counted.

    Used for **reading**, never for deciding: the drawer sorts a layer's rules
    broadest-first so the bound that covers everything is read before the
    exceptions to it. Most-specific-wins is the rule this product does not have.
    """
    return sum(1 for segment in pattern.strip().split("/") if "*" not in segment)

File: studio/src/pages/graphs-detail/features/govern/addressing.ts
/**
 * An address, taken apart and put back together — so every control in the
 * authoring form is a **pick** rather than a field somebody types into
 * ([WO7](../../../../../docs/for-developers/modules/govern/features/worlds.md)).
 *
 * `<layer>/<sublayer>/<name>` is the only identifier a rule has
 * ([GV4](../../../../../docs/for-developers/modules/govern/spec.md)), and the
 * three wildcards are part of the vocabulary rather than an escape from it:
 * `*` is one segment, `**` is the rest, and a `*` **inside** a segment globs it
 * — which is how `graph_data/model/Deals@*` names a model across its versions
 * without stopping the day somebody publishes a new one.
 *
 * **The wildcards are options, not typing.** Offering `Deals@*` beside
 * `Deals@1.0.1` in the same picker is what makes the version-proof rule as easy
 * to write as the brittle one; leaving it to a keystroke makes the brittle one
 * the default.
 */

import type { GovernLayer, Participant } from "@/types/govern";

/** Any sublayer. One segment — the middle one. */
export const ANY_SUBLAYER = "*";
/** Everything under here. The rest of the address, however many segments. */
export const REST = "**";

export interface AddressParts {
	layer: GovernLayer;
	sublayer: string;
	name: string;
}

/**
 * Three picks → one pattern.
 *
 * `graph_data/*\/**` collapses to `graph_data/**`: a middle segment that
 * matches anything, followed by a tail that matches everything, says nothing
 * the tail does not already say — and two spellings of one bound are two things
 * an auditor has to know are the same.
 */
export function buildMatch({ layer, sublayer, name }: AddressParts): string {
	if (sublayer === ANY_SUBLAYER && name === REST) return `${layer}/${REST}`;
	return `${layer}/${sublayer}/${name}`;
}

/** A pattern → the three picks that would rebuild it. */
export function splitMatch(match: string): AddressParts {
	const [layer, ...rest] = match.split("/");
	if (!rest.length) {
		return { layer: layer as GovernLayer, sublayer: ANY_SUBLAYER, name: REST };
	}
	if (rest.length === 1) {
		return {
			layer: layer as GovernLayer,
			sublayer: rest[0] === REST ? ANY_SUBLAYER : rest[0],
			name: REST,
		};
	}
	// A third-party address runs past three segments — `third_party/api/
	// clearbit.com/v2/companies` — so the name is everything after the sublayer.
	return {
		layer: layer as GovernLayer,
		sublayer: rest[0],
		name: rest.slice(1).join("/"),
	};
}

/** The sublayers this layer actually has, in the order the catalogue gives them. */
export function sublayersIn(
	participants: Participant[],
	layer: GovernLayer,
): string[] {
	return [
		...new Set(
			participants.filter((p) => p.layer === layer).map((p) => p.sublayer),
		),
	];
}

/**
 * What the name picker offers for one (layer, sublayer).
 *
 * Every participant by its exact name, plus — for anything carrying an `@`
 * discriminator — the `@*` form beside it. A model version and *every version
 * of that model* are two different bounds, and only one of them survives the
 * next publish.
 */
export function nameOptionsIn(
	participants: Participant[],
	layer: GovernLayer,
	sublayer: string,
): { value: string; label: string; description?: string }[] {
	const here = participants.filter(
		(p) =>
			p.layer === layer &&
			(sublayer === ANY_SUBLAYER || p.sublayer === sublayer),
	);

	const exact = here.map((p) => ({
		value: p.name,
		label: p.name,
		description: p.label,
	}));

	const globbed = [
		...new Set(
			here
				.filter((p) => p.name.includes("@"))
				.map((p) => `${p.name.split("@")[0]}@*`),
		),
	].map((value) => ({
		value,
		label: value,
		description: "every published version — survives the next publish",
	}));

	return [
		{
			value: REST,
			label: REST,
			description: "everything under here, however deep",
		},
		...globbed,
		...exact,
	];
}

/**
 * Does this pattern reach this address? The same three wildcards the engine
 * reads, so the preview and the save never disagree about what a rule bites.
 *
 * Mirrors `apps/govern/addressing.py`. It is duplicated rather than asked for
 * because the preview answers on every keystroke of a pick, and a round trip
 * per pick would make the near-misses arrive after the eye has moved on — the
 * catalogue itself is still the server's (GV21), only the matching is local.
 */
export function matches(pattern: string, address: string): boolean {
	const p = pattern.split("/");
	const a = address.split("/");

	for (let i = 0; i < p.length; i++) {
		if (p[i] === REST) return true;
		if (i >= a.length) return false;
		if (!segmentMatches(p[i], a[i])) return false;
	}
	return p.length === a.length;
}

function segmentMatches(pattern: string, segment: string): boolean {
	if (pattern === ANY_SUBLAYER) return true;
	if (!pattern.includes(ANY_SUBLAYER)) return pattern === segment;
	// A `*` inside a segment globs it — `Deals@*`, `claude-*`.
	const escaped = pattern
		.split(ANY_SUBLAYER)
		.map((part) => part.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"))
		.join(".*");
	return new RegExp(`^${escaped}$`).test(segment);
}
